### PR TITLE
Use ARM metadata to configure clouds (endpoints/suffixes, etc.)

### PR DESCRIPTION
**Pull request by mdewan**
---
In this change, we attempt to load cloud metadata info from ARM endpoints as opposed to using hard coded values in the CLI source. If for whatever reason, we fail to retrieve or use the ARM metadata info, we fall back on the hard coded values. 

---
This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
@Juliehzl please review the PR.

**Comment by Juliehzl** [comment]
Hi @mdewan, can you try to fix CI first?

**Comment by mdewan** [comment]
@Juliehzl , CI is green now

> Hi @mdewan, can you try to fix CI first?

File: src/azure-cli-core/azure/cli/core/cloud.py [hunk 1]
```diff
@@ -17,6 +22,7 @@
 logger = get_logger(__name__)
 
 CLOUD_CONFIG_FILE = os.path.join(GLOBAL_CONFIG_DIR, 'clouds.config')
+ARM_CLOUD_METADATA_URL = 'https://management.azure.com/metadata/endpoints?api-version=2019-05-01'
```
Comment (Juliehzl) [inline src/azure-cli-core/azure/cli/core/cloud.py]: The request url is fixed to api-version 2019-05-01?

File: src/azure-cli-core/azure/cli/core/cloud.py [hunk 2]
```diff
@@ -127,6 +133,78 @@ def __getattribute__(self, name):
...
+
+
+def urlretrieve(url):
+    req = urlopen(url, context=_ssl_context())
+    return req.read()
+
+
+def get_ossrdbms_resource_id(cloud_name):
+    ossrdbms_mapper = {
+        'AzureCloud': 'https://ossrdbms-aad.database.windows.net',
```
Comment (Juliehzl) [inline src/azure-cli-core/azure/cli/core/cloud.py]: What is ossrdbms?

File: src/azure-cli-core/azure/cli/core/cloud.py [hunk 3]
```diff
@@ -127,6 +133,78 @@ def __getattribute__(self, name):
...
+            return ssl.SSLContext(ssl.PROTOCOL_TLS)  # added in python 2.7.13 and 3.6
+        except AttributeError:
+            return ssl.SSLContext(ssl.PROTOCOL_TLSv1)
+
+    return ssl.create_default_context()
+
+
+def urlretrieve(url):
+    req = urlopen(url, context=_ssl_context())
+    return req.read()
```
Comment (yugangw-msft) [inline src/azure-cli-core/azure/cli/core/cloud.py]: This is duplication code copied and pasted from other place, please refactor

File: src/azure-cli-core/azure/cli/core/cloud.py (hunk 1 above)
Comment (yugangw-msft) [reply src/azure-cli-core/azure/cli/core/cloud.py]: Is this endpoint available in national cloud? If not, then we should only configure public cloud and leave the existing static national cloud alone

Comment (mdewan) [reply src/azure-cli-core/azure/cli/core/cloud.py]: This endpoint is not available in all clouds. I'm going to change the logic a bit, and update this PR.

Comment (mdewan) [reply src/azure-cli-core/azure/cli/core/cloud.py]: Updated the PR, and took care of this using an environment variable

File: src/azure-cli-core/azure/cli/core/cloud.py (hunk 2 above)
Comment (mdewan) [reply src/azure-cli-core/azure/cli/core/cloud.py]: OSS relational DB offerings in Azure. A blog at https://azure.microsoft.com/en-us/blog/happy-birthday-to-managed-open-source-rdbms-services-in-azure/ has some context.

OSS RDBMS config currently exists in CLI, so there's nothing new here.

File: src/azure-cli-core/azure/cli/core/cloud.py (hunk 3 above)
Comment (mdewan) [reply src/azure-cli-core/azure/cli/core/cloud.py]: Done, moved to util.py

**Comment by mdewan** [comment]
> Other than the review comments, we will only accept this PR if all service teams agree that every ever new endpoints will be exposed from the this endpoint.
> 
> Please let @achandmsft coordinate

We have been working with @achandmsft, and he is aware and supportive of this work and pattern.

File: src/azure-cli-core/azure/cli/core/cloud.py [hunk 4]
```diff
@@ -153,94 +208,100 @@ def __str__(self):
...
-
-
-KNOWN_CLOUDS = [AZURE_PUBLIC_CLOUD, AZURE_CHINA_CLOUD, AZURE_US_GOV_CLOUD, AZURE_GERMAN_CLOUD]
+try:
+    arm_cloud_dict = json.loads(urlretrieve(os.getenv('ARM_CLOUD_METADATA_URL')))
+    cli_cloud_dict = convert_arm_to_cli(arm_cloud_dict)
+    if 'AzureCloud' in cli_cloud_dict:
+        cli_cloud_dict['AzureCloud'].endpoints.active_directory = 'https://login.microsoftonline.com'  # pylint: disable=line-too-long # change once active_directory is fixed in ARM for the public cloud
+    KNOWN_CLOUDS = list(cli_cloud_dict.values())
+except Exception:  # pylint: disable=broad-except
```
Comment (yugangw-msft) [inline src/azure-cli-core/azure/cli/core/cloud.py]: I will suggest we remove the try~catch and keep the hard coded ones; then at the end, detect the env var is set, then load from there.

Comment (mdewan) [reply src/azure-cli-core/azure/cli/core/cloud.py]: makes sense, done

File: src/azure-cli-core/azure/cli/core/cloud.py [hunk 5]
```diff
@@ -239,10 +294,21 @@ def __str__(self):
...
 
+try:
+    if 'ARM_CLOUD_METADATA_URL' in os.environ:
+        arm_cloud_dict = json.loads(urlretrieve(os.getenv('ARM_CLOUD_METADATA_URL')))
+        cli_cloud_dict = convert_arm_to_cli(arm_cloud_dict)
+        if 'AzureCloud' in cli_cloud_dict:
+            cli_cloud_dict['AzureCloud'].endpoints.active_directory = 'https://login.microsoftonline.com'  # pylint: disable=line-too-long # change once active_directory is fixed in ARM for the public cloud
+        KNOWN_CLOUDS = list(cli_cloud_dict.values())
+except Exception as ex:  # pylint: disable=broad-except
+    print('The ARM_CLOUD_METADATA_URL environment variable points to an invalid cloud metadata URL')
```
Comment (yugangw-msft) [inline src/azure-cli-core/azure/cli/core/cloud.py]: please use `logger.warn('Failed to load cloud metadata from the url specified by ARM_CLOUD_METADATA_URL')`

File: src/azure-cli-core/azure/cli/core/cloud.py [hunk 6]
```diff
@@ -127,6 +129,59 @@ def __getattribute__(self, name):
         return val
 
 
+def get_ossrdbms_resource_id(cloud_name):
```
Comment (yugangw-msft) [inline src/azure-cli-core/azure/cli/core/cloud.py]: please start with "_" for all methods where are private to the current module, and do it across

File: src/azure-cli-core/azure/cli/core/cloud.py [hunk 7]
```diff
@@ -239,10 +294,21 @@ def __str__(self):
         keyvault_dns='.vault.microsoftazure.de',
         sql_server_hostname='.database.cloudapi.de'))
 
-
 KNOWN_CLOUDS = [AZURE_PUBLIC_CLOUD, AZURE_CHINA_CLOUD, AZURE_US_GOV_CLOUD, AZURE_GERMAN_CLOUD]
 
 
+try:
+    if 'ARM_CLOUD_METADATA_URL' in os.environ:
```
Comment (yugangw-msft) [inline src/azure-cli-core/azure/cli/core/cloud.py]: Move this out of the `try` block

Comment (mdewan) [reply src/azure-cli-core/azure/cli/core/cloud.py]: done

File: src/azure-cli-core/azure/cli/core/cloud.py (hunk 6 above)
Comment (mdewan) [reply src/azure-cli-core/azure/cli/core/cloud.py]: done, thanks

File: src/azure-cli-core/azure/cli/core/cloud.py (hunk 5 above)
Comment (mdewan) [reply src/azure-cli-core/azure/cli/core/cloud.py]: done

**Comment by yugangw-msft** [comment]
@Juliehzl, please encourage your colleagues test it

**Pull request merged**